lib/src/annotate.rs

```diff
@@ -0,0 +1,374 @@
+    /// For each value in the array, the commit_id is the commit id of the
+    /// originator of the line and the string is the actual line itself (without
+    /// newline terminators). The vector is ordered by appearance in the
+    /// file
+    pub file_annotations: Vec<(CommitId, Vec<u8>)>,
+}
+
+/// A map from commits to line mappings.
+/// Namely, for a given commit A, the value is the mapping of lines in the file
+/// at commit A to line numbers in the original file
+type CommitLineMap = HashMap<CommitId, HashMap<usize, usize>>;
+
+/// Memoizes the file contents for a given version to save time
+type FileCache = HashMap<CommitId, Vec<u8>>;
+
+/// A map from line numbers in the original file to the commit that originated
+/// that line
+type OriginalLineMap = HashMap<usize, CommitId>;
+
+fn get_initial_commit_line_map(commit_id: &CommitId, num_lines: usize) -> CommitLineMap {
+    let mut starting_commit_map = HashMap::new();
+    for i in 0..num_lines {
+        starting_commit_map.insert(i, i);
+    }
+
+    let mut starting_line_map = HashMap::new();
+    starting_line_map.insert(commit_id.clone(), starting_commit_map);
+    starting_line_map
+}
+
+/// Once we've looked at all parents of a commit, any leftover lines must be
+/// original to the current commit, so we save this information in
+/// original_line_map.
+fn mark_lines_from_original(
+    original_line_map: &mut OriginalLineMap,
+    commit_id: &CommitId,
+    commit_lines: HashMap<usize, usize>,
+) {
+    for (_, original_line_number) in commit_lines {
+        original_line_map.insert(original_line_number, commit_id.clone());
+    }
+}
+
+/// Takes in an original line map and the original contents and annotates each
+/// line according to the contents of the provided OriginalLineMap
+fn convert_to_results(
+    original_line_map: OriginalLineMap,
+    original_contents: &[u8],
+) -> AnnotateResults {
+    let mut result_lines = Vec::new();
+    original_contents
+        .split_inclusive(|b| *b == b'\n')
+        .enumerate()
+        .for_each(|(idx, line)| {
+            result_lines.push((
+                original_line_map.get(&idx).unwrap().clone(),
+                line.to_owned(),
+            ));
+        });
+    AnnotateResults {
+        file_annotations: result_lines,
+    }
+}
+
+/// loads a given file into the cache under a specific commit id.
+/// If there is already a file for a given commit, it is a no-op.
+fn load_file_into_cache(
+    file_cache: &mut FileCache,
+    store: &Store,
+    commit_id: &CommitId,
+    file_path: &RepoPath,
+    tree: &MergedTree,
+) -> Result<(), BackendError> {
+    if file_cache.contains_key(commit_id) {
+        return Ok(());
+    }
+
+    if let Some(file_contents) = get_file_contents(store, file_path, tree)? {
+        file_cache.insert(commit_id.clone(), file_contents);
+    } else {
+        // file is missing or is not a regular file (e.g. directory)
+        file_cache.insert(commit_id.clone(), Vec::new());
```

Comment:
nit: (at your option) `get_file_contents()` can be changed to return `Vec<u8>` instead of `Option<_>`. An absent/non-file entry is equivalent to an empty file during annotation computation.